File: android/app/src/main/java/io/github/magnusencoded/setlist2spotify/data/spotify/SpotifyClient.kt
package io.github.magnusencoded.setlist2spotify.data.spotify

import android.net.Uri
import android.util.Base64
import io.github.magnusencoded.setlist2spotify.data.SettingsRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import okhttp3.FormBody
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.security.MessageDigest
import java.security.SecureRandom

const val SPOTIFY_REDIRECT_URI = "setlist2spotify://callback"
private const val SPOTIFY_SCOPES =
    "playlist-modify-public playlist-modify-private user-read-private"

class SpotifyClient(private val settings: SettingsRepository) {

    private val http = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    // --- OAuth (Authorization Code with PKCE) ---

    suspend fun buildAuthorizationUri(): Uri {
        val clientId = settings.spotifyClientIdValue()
            ?: throw IOException("Spotify Client ID is not configured. Set it in Settings.")
        val verifier = generateCodeVerifier()
        settings.savePkceVerifier(verifier)
        return Uri.parse("https://accounts.spotify.com/authorize").buildUpon()
            .appendQueryParameter("client_id", clientId)
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("redirect_uri", SPOTIFY_REDIRECT_URI)
            .appendQueryParameter("code_challenge_method", "S256")
            .appendQueryParameter("code_challenge", codeChallenge(verifier))
            .appendQueryParameter("scope", SPOTIFY_SCOPES)
            // Always show the consent screen so a stale earlier grant without
            // playlist scopes can't be silently reused.
            .appendQueryParameter("show_dialog", "true")
            .build()
    }

    suspend fun exchangeCodeForTokens(code: String) {
        val clientId = settings.spotifyClientIdValue()
            ?: throw IOException("Spotify Client ID is not configured.")
        val verifier = settings.pkceVerifier()
            ?: throw IOException("Login session expired. Start the Spotify login again.")
        val body = FormBody.Builder()
            .add("grant_type", "authorization_code")
            .add("code", code)
            .add("redirect_uri", SPOTIFY_REDIRECT_URI)
            .add("client_id", clientId)
            .add("code_verifier", verifier)
            .build()
        val token = requestToken(body)
        settings.saveTokens(token.accessToken, token.refreshToken, token.expiresIn, token.scope)
    }

    private suspend fun refreshAccessToken(): String {
        val clientId = settings.spotifyClientIdValue()
            ?: throw IOException("Spotify Client ID is not configured.")
        val refreshToken = settings.refreshTokenValue()
            ?: throw IOException("Not connected to Spotify. Connect in Settings.")
        val body = FormBody.Builder()
            .add("grant_type", "refresh_token")
            .add("refresh_token", refreshToken)
            .add("client_id", clientId)
            .build()
        val token = requestToken(body)
        settings.saveTokens(token.accessToken, token.refreshToken, token.expiresIn, token.scope)
        return token.accessToken
    }

    /** Null when unknown (logins predating scope persistence). */
    suspend fun hasPlaylistScopes(): Boolean? =
        settings.grantedScope()?.contains("playlist-modify")

    private suspend fun requestToken(body: FormBody): TokenResponse = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("https://accounts.spotify.com/api/token")
            .post(body)
            .build()
        http.newCall(request).execute().use { resp ->
            val text = resp.body?.string() ?: ""
            if (!resp.isSuccessful) {
                throw IOException("Spotify token request failed (${resp.code}): $text")
            }
            json.decodeFromString<TokenResponse>(text)
        }
    }

    private suspend fun accessToken(): String =
        settings.validAccessToken() ?: refreshAccessToken()

    suspend fun isConnected(): Boolean = settings.refreshTokenValue() != null

    // --- Web API ---

    private suspend fun call(request: Request.Builder): String {
        val token = accessToken()
        val req = request.header("Authorization", "Bearer $token").build()
        return withContext(Dispatchers.IO) {
            http.newCall(req).execute().use { resp ->
                val text = resp.body?.string() ?: ""
                if (!resp.isSuccessful) {
                    if (resp.code == 401) {
                        throw IOException("Spotify session expired. Reconnect in Settings.")
                    }
                    if (resp.code == 403) {
                        // Spotify's bare "Forbidden" body says nothing; the headers
                        // distinguish a scope refusal from an edge/CDN block.
                        val detail = listOf("www-authenticate", "server", "x-robots-tag")
                            .mapNotNull { name -> resp.header(name)?.let { "$name: $it" } }
                            .joinToString("; ")
                        throw SpotifyForbiddenException(
                            "Spotify refused ${req.method} ${req.url.encodedPath} (403). " +
                                "$text${if (detail.isNotEmpty()) " [$detail]" else ""}"
                        )
                    }
                    throw IOException(
                        "Spotify API error ${resp.code} on ${req.method} ${req.url.encodedPath}: $text"
                    )
                }
                text
            }
        }
    }

    suspend fun searchTracks(query: String, limit: Int = 5): List<SpotifyTrack> {
        val url = "https://api.spotify.com/v1/search".toHttpUrl().newBuilder()
            .addQueryParameter("q", query)
            .addQueryParameter("type", "track")
            .addQueryParameter("limit", limit.toString())
            .build()
        val text = call(Request.Builder().url(url))
        return json.decodeFromString<TrackSearchResponse>(text).tracks?.items.orEmpty()
    }

    suspend fun createPlaylist(name: String, description: String, isPublic: Boolean): PlaylistResponse {
        val payload = buildJsonObject {
            put("name", name)
            put("description", description)
            put("public", isPublic)
        }
        // /me/playlists avoids the user-id round trip and the 403s the
        // /users/{id}/playlists endpoint gives on any id mismatch.
        val request = Request.Builder()
            .url("https://api.spotify.com/v1/me/playlists")
            .post(payload.toString().toRequestBody(jsonMediaType))
        return json.decodeFromString(call(request))
    }

    suspend fun currentUser(): SpotifyUser =
        json.decodeFromString(call(Request.Builder().url("https://api.spotify.com/v1/me")))

    /** Reads a playlist by id — used to harvest the creator's setlist.fm stamp from a shared link. */
    suspend fun getPlaylist(playlistId: String): SimplePlaylist {
        val url = "https://api.spotify.com/v1/playlists/$playlistId".toHttpUrl().newBuilder()
            .addQueryParameter("fields", "id,name,description,owner(id,display_name)")
            .build()
        return json.decodeFromString(call(Request.Builder().url(url)))
    }

    private fun urisBody(uris: List<String>) =
        buildJsonObject { putJsonArray("uris") { uris.forEach { add(it) } } }
            .toString()
            .toRequestBody(jsonMediaType)

    private suspend fun makePlaylistPublic(playlistId: String) {
        val body = buildJsonObject { put("public", true) }.toString().toRequestBody(jsonMediaType)
        call(Request.Builder().url("https://api.spotify.com/v1/playlists/$playlistId").put(body))
    }

    /**
     * Facts that identify why Spotify refuses to modify a playlist we just made.
     * [detailsWrite] records whether a non-item write to the same playlist was
     * accepted, which separates "this app cannot add items" from "this app
     * cannot write playlists at all".
     */
    private suspend fun diagnostics(playlistId: String, detailsWrite: String): String = try {
        val me = currentUser()
        val playlist = call(
            Request.Builder()
                .url("https://api.spotify.com/v1/playlists/$playlistId?fields=owner(id),public,collaborative")
        )
        "me=${me.id} product=${me.product} playlist=$playlist " +
            "detailsWrite=$detailsWrite scopes=${settings.grantedScope()}"
    } catch (e: Exception) {
        "diagnostics unavailable: ${e.message}"
    }

    /**
     * Fills a freshly created playlist. Spotify has been refusing the documented
     * call with a bare 403 that carries no www-authenticate header, so every
     * accepted way of expressing "add these tracks" is tried before giving up,
     * and the final failure carries the facts needed to explain it.
     */
    suspend fun addTracks(playlistId: String, uris: List<String>): AddTracksResult {
        val clean = uris.filter { it.startsWith("spotify:track:") }.distinct()
        if (clean.isEmpty()) throw IOException("No valid Spotify track URIs to add.")
        val url = "https://api.spotify.com/v1/playlists/$playlistId/tracks"

        // Over 100 items only the batched form is safe: the alternatives below
        // either replace the contents or would leave a half-filled playlist.
        if (clean.size > 100) {
            clean.chunked(100).forEach { call(Request.Builder().url(url).post(urisBody(it))) }
            return AddTracksResult(clean.size, emptyList())
        }

        var lastError: SpotifyForbiddenException? = null

        suspend fun attempt(build: () -> Request.Builder): AddTracksResult? = try {
            call(build())
            AddTracksResult(clean.size, emptyList())
        } catch (e: SpotifyForbiddenException) {
            lastError = e
            null
        }

        // 1. The documented form: uris in the JSON body.
        attempt { Request.Builder().url(url).post(urisBody(clean)) }?.let { return it }

        // 2. PUT sets the contents of a playlist that is still empty.
        attempt { Request.Builder().url(url).put(urisBody(clean)) }?.let { return it }

        // 3. The other documented form: uris as a query parameter.
        val queryUrl = url.toHttpUrl().newBuilder()
            .addQueryParameter("uris", clean.joinToString(","))
            .build()
        attempt {
            Request.Builder().url(queryUrl).post("".toRequestBody(jsonMediaType))
        }?.let { return it }

        // 4. A grant may cover public playlists only, so try again as public.
        var detailsWrite = "ok"
        try {
            makePlaylistPublic(playlistId)
        } catch (e: SpotifyForbiddenException) {
            detailsWrite = "refused"
            lastError = e
        }
        if (detailsWrite == "ok") {
            attempt { Request.Builder().url(url).post(urisBody(clean)) }?.let { return it }
        }

        // 5. One at a time, so a single refused track cannot cost the rest.
        var added = 0
        val refused = mutableListOf<String>()
        for (uri in clean) {
            try {
                call(Request.Builder().url(url).post(urisBody(listOf(uri))))
                added++
            } catch (e: SpotifyForbiddenException) {
                refused += uri
                lastError = e
            }
        }
        if (added > 0) return AddTracksResult(added, refused)

        throw SpotifyForbiddenException(
            "${lastError?.message} | ${diagnostics(playlistId, detailsWrite)}"
        )
    }
}

data class AddTracksResult(val added: Int, val refused: List<String>)

class SpotifyForbiddenException(message: String) : IOException(message)

private fun generateCodeVerifier(): String {
    val bytes = ByteArray(64)
    SecureRandom().nextBytes(bytes)
    return Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
}

private fun codeChallenge(verifier: String): String {
    val digest = MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray(Charsets.US_ASCII))
    return Base64.encodeToString(digest, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
}

File: ios/StationToStation/Data/Spotify/SpotifyClient.swift
import Foundation
import CryptoKit
import AuthenticationServices
import UIKit

let spotifyRedirectURI = "station-to-station://callback"
private let spotifyScopes = "playlist-modify-public playlist-modify-private user-read-private ugc-image-upload"

struct AddTracksResult { let added: Int; let refused: [String] }

final class SpotifyClient {

    private let settings: Settings
    private let decoder = JSONDecoder()
    private var authSession: ASWebAuthenticationSession?
    private let authAnchor = AuthAnchor()

    init(_ settings: Settings) { self.settings = settings }

    // --- OAuth (Authorization Code with PKCE) ---
    //
    // Android launches a browser and catches the station-to-station://callback deep
    // link. iOS has ASWebAuthenticationSession, which runs the whole round trip
    // and hands back the callback URL directly — no URL-scheme plumbing for the
    // redirect, and the sheet auto-dismisses.

    @MainActor
    func login() async throws {
        guard let clientId = settings.spotifyClientIdValue else {
            throw AppError("Spotify Client ID is not configured. Set it in Settings.")
        }
        let verifier = generateCodeVerifier()
        settings.savePkceVerifier(verifier)
        var comps = URLComponents(string: "https://accounts.spotify.com/authorize")!
        comps.queryItems = [
            .init(name: "client_id", value: clientId),
            .init(name: "response_type", value: "code"),
            .init(name: "redirect_uri", value: spotifyRedirectURI),
            .init(name: "code_challenge_method", value: "S256"),
            .init(name: "code_challenge", value: codeChallenge(verifier)),
            .init(name: "scope", value: spotifyScopes),
            // Always show consent so a stale grant without playlist scopes can't
            // be silently reused.
            .init(name: "show_dialog", value: "true"),
        ]
        let callback = try await authenticate(url: comps.url!)
        let cc = URLComponents(url: callback, resolvingAgainstBaseURL: false)
        if let error = cc?.queryItems?.first(where: { $0.name == "error" })?.value {
            throw AppError("Spotify login failed: \(error)")
        }
        guard let code = cc?.queryItems?.first(where: { $0.name == "code" })?.value else {
            throw AppError("Spotify login returned no authorization code.")
        }
        try await exchangeCodeForTokens(code)
    }

    @MainActor
    private func authenticate(url: URL) async throws -> URL {
        try await withCheckedThrowingContinuation { continuation in
            let session = ASWebAuthenticationSession(
                url: url, callbackURLScheme: "station-to-station"
            ) { callbackURL, error in
                if let callbackURL {
                    continuation.resume(returning: callbackURL)
                } else if let error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume(throwing: AppError("Spotify login was cancelled."))
                }
            }
            session.presentationContextProvider = authAnchor
            session.prefersEphemeralWebBrowserSession = false
            authSession = session
            session.start()
        }
    }

    private func exchangeCodeForTokens(_ code: String) async throws {
        guard let clientId = settings.spotifyClientIdValue else {
            throw AppError("Spotify Client ID is not configured.")
        }
        guard let verifier = settings.pkceVerifier else {
            throw AppError("Login session expired. Start the Spotify login again.")
        }
        let token = try await requestToken([
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": spotifyRedirectURI,
            "client_id": clientId,
            "code_verifier": verifier,
        ])
        settings.saveTokens(access: token.accessToken, refresh: token.refreshToken,
                            expiresIn: token.expiresIn, scope: token.scope)
    }

    private func refreshAccessToken() async throws -> String {
        guard let clientId = settings.spotifyClientIdValue else {
            throw AppError("Spotify Client ID is not configured.")
        }
        guard let refresh = settings.refreshTokenValue else {
            throw AppError("Not connected to Spotify. Connect in Settings.")
        }
        let token = try await requestToken([
            "grant_type": "refresh_token",
            "refresh_token": refresh,
            "client_id": clientId,
        ])
        settings.saveTokens(access: token.accessToken, refresh: token.refreshToken,
                            expiresIn: token.expiresIn, scope: token.scope)
        return token.accessToken
    }

    /// Nil when unknown (logins predating scope persistence).
    func hasPlaylistScopes() -> Bool? {
        settings.grantedScope.map { $0.contains("playlist-modify") }
    }

    /// Cover upload needs a scope the app did not always ask for, so a login
    /// made before covers existed can create playlists but not illustrate them.
    func hasImageUploadScope() -> Bool {
        settings.grantedScope?.contains("ugc-image-upload") == true
    }

    private func requestToken(_ fields: [String: String]) async throws -> TokenResponse {
        var request = URLRequest(url: URL(string: "https://accounts.spotify.com/api/token")!)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = formEncode(fields).data(using: .utf8)
        let (data, response) = try await URLSession.shared.data(for: request)
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200...299).contains(code) else {
            throw AppError("Spotify token request failed (\(code)): \(String(data: data, encoding: .utf8) ?? "")")
        }
        return try decoder.decode(TokenResponse.self, from: data)
    }

    private func accessToken() async throws -> String {
        if let token = settings.validAccessToken { return token }
        return try await refreshAccessToken()
    }

    var isConnected: Bool { settings.refreshTokenValue != nil }

    // --- Web API ---

    private func call(_ build: (inout URLRequest) -> Void) async throws -> Data {
        let token = try await accessToken()
        var request = URLRequest(url: URL(string: "https://api.spotify.com")!)
        build(&request)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (data, response) = try await URLSession.shared.data(for: request)
        let http = response as? HTTPURLResponse
        let code = http?.statusCode ?? 0
        let path = request.url?.path ?? ""
        let method = request.httpMethod ?? "GET"
        guard (200...299).contains(code) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            if code == 401 {
                throw AppError("Spotify session expired. Reconnect in Settings.")
            }
            if code == 403 {
                // Spotify's bare "Forbidden" body says nothing; the headers
                // distinguish a scope refusal from an edge/CDN block.
                let detail = ["www-authenticate", "server", "x-robots-tag"]
                    .compactMap { name in (http?.value(forHTTPHeaderField: name)).map { "\(name): \($0)" } }
                    .joined(separator: "; ")
                throw SpotifyForbidden(
                    "Spotify refused \(method) \(path) (403). \(body)"
                        + (detail.isEmpty ? "" : " [\(detail)]"))
            }
            throw AppError("Spotify API error \(code) on \(method) \(path): \(body)")
        }
        return data
    }

    func searchTracks(_ query: String, limit: Int = 5) async throws -> [SpotifyTrack] {
        var comps = URLComponents(string: "https://api.spotify.com/v1/search")!
        comps.queryItems = [
            .init(name: "q", value: query),
            .init(name: "type", value: "track"),
            .init(name: "limit", value: "\(limit)"),
        ]
        let url = comps.url!
        let data = try await call { $0.url = url }
        return try decoder.decode(TrackSearchResponse.self, from: data).tracks?.items ?? []
    }

    func createPlaylist(name: String, description: String, isPublic: Bool) async throws -> PlaylistResponse {
        let payload: [String: Any] = ["name": name, "description": description, "public": isPublic]
        let body = try JSONSerialization.data(withJSONObject: payload)
        // /me/playlists avoids the user-id round trip and the 403s the
        // /users/{id}/playlists endpoint gives on any id mismatch.
        let data = try await call {
            $0.url = URL(string: "https://api.spotify.com/v1/me/playlists")!
            $0.httpMethod = "POST"
            $0.setValue("application/json", forHTTPHeaderField: "Content-Type")
            $0.httpBody = body
        }
        return try decoder.decode(PlaylistResponse.self, from: data)
    }

    func currentUser() async throws -> SpotifyUser {
        let data = try await call { $0.url = URL(string: "https://api.spotify.com/v1/me")! }
        return try decoder.decode(SpotifyUser.self, from: data)
    }

    /// Reads a playlist by id — used to harvest the creator's setlist.fm stamp from a shared link.
    func getPlaylist(_ playlistId: String) async throws -> SimplePlaylist {
        var comps = URLComponents(string: "https://api.spotify.com/v1/playlists/\(playlistId)")!
        comps.queryItems = [.init(name: "fields", value: "id,name,description,owner(id,display_name)")]
        let url = comps.url!
        let data = try await call { $0.url = url }
        return try decoder.decode(SimplePlaylist.self, from: data)
    }

    /// Facts that explain a residual 403 while modifying a playlist we just made.
    private func diagnostics(_ playlistId: String) async -> String {
        do {
            let me = try await currentUser()
            let data = try await call {
                $0.url = URL(string: "https://api.spotify.com/v1/playlists/\(playlistId)?fields=owner(id),public,collaborative")!
            }
            let playlist = String(data: data, encoding: .utf8) ?? ""
            return "me=\(me.id) product=\(me.product ?? "?") playlist=\(playlist) scopes=\(settings.grantedScope ?? "nil")"
        } catch {
            return "diagnostics unavailable: \(userMessage(error))"
        }
    }

    /// Fills a freshly created playlist via POST /playlists/{id}/items. The old
    /// /tracks path was renamed in Spotify's Feb 2026 API migration and now 403s
    /// from the edge (envoy, no www-authenticate); /items takes the same body.
    func addTracks(_ playlistId: String, uris: [String]) async throws -> AddTracksResult {
        var seen = Set<String>()
        let clean = uris.filter { $0.hasPrefix("spotify:track:") && seen.insert($0).inserted }
        if clean.isEmpty { throw AppError("No valid Spotify track URIs to add.") }
        let url = URL(string: "https://api.spotify.com/v1/playlists/\(playlistId)/items")!
        do {
            for chunk in clean.chunked(into: 100) {
                let body = try JSONSerialization.data(withJSONObject: ["uris": chunk])
                _ = try await call {
                    $0.url = url
                    $0.httpMethod = "POST"
                    $0.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    $0.httpBody = body
                }
            }
        } catch let e as SpotifyForbidden {
            throw SpotifyForbidden("\(e.message) | \(await diagnostics(playlistId))")
        }
        return AddTracksResult(added: clean.count, refused: [])
    }

    /// Sets the playlist cover. Spotify takes the JPEG base64-encoded as the raw
    /// body under an image/jpeg content type — not multipart, and not wrapped in
    /// JSON — and answers 202 with nothing in the body.
    func uploadCover(_ playlistId: String, jpeg: Data) async throws {
        let body = jpeg.base64EncodedData()
        _ = try await call {
            $0.url = URL(string: "https://api.spotify.com/v1/playlists/\(playlistId)/images")!
            $0.httpMethod = "PUT"
            $0.setValue("image/jpeg", forHTTPHeaderField: "Content-Type")
            $0.httpBody = body
        }
    }
}

// --- PKCE helpers ---

private func generateCodeVerifier() -> String {
    var bytes = [UInt8](repeating: 0, count: 64)
    _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
    return base64URL(Data(bytes))
}

private func codeChallenge(_ verifier: String) -> String {
    base64URL(Data(SHA256.hash(data: Data(verifier.utf8))))
}

private func base64URL(_ data: Data) -> String {
    data.base64EncodedString()
        .replacingOccurrences(of: "+", with: "-")
        .replacingOccurrences(of: "/", with: "_")
        .replacingOccurrences(of: "=", with: "")
}

private func formEncode(_ fields: [String: String]) -> String {
    var allowed = CharacterSet.alphanumerics
    allowed.insert(charactersIn: "-._~")
    return fields.map { k, v in
        "\(k)=\(v.addingPercentEncoding(withAllowedCharacters: allowed) ?? v)"
    }.joined(separator: "&")
}

/// Presentation anchor for ASWebAuthenticationSession.
private final class AuthAnchor: NSObject, ASWebAuthenticationPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        UIApplication.shared.connectedScenes
            .compactMap { ($0 as? UIWindowScene)?.keyWindow }
            .first ?? ASPresentationAnchor()
    }
}

extension Array {
    func chunked(into size: Int) -> [[Element]] {
        stride(from: 0, to: count, by: size).map { Array(self[$0 ..< Swift.min($0 + size, count)]) }
    }
}
